File: app/src/main/java/com/raju/disney/opentelemetry/ActiveSpan.java
package com.raju.disney.opentelemetry;

import java.util.function.Supplier;

import io.opentelemetry.api.trace.Span;
import io.opentelemetry.context.Scope;

class ActiveSpan {
    private final VisibleScreenTracker visibleScreenTracker;

    private Span span;
    private Scope scope;

    ActiveSpan(VisibleScreenTracker visibleScreenTracker) {
        this.visibleScreenTracker = visibleScreenTracker;
    }

    boolean spanInProgress() {
        return span != null;
    }

    void startSpan(Supplier<Span> spanCreator) {
        //don't start one if there's already one in progress
        if (span != null) {
            return;
        }
        this.span = spanCreator.get();
        scope = span.makeCurrent();
    }

    void endActiveSpan() {
        if (scope != null) {
            scope.close();
            scope = null;
        }
        if (this.span != null) {
            this.span.end();
            this.span = null;
        }
    }

    void addEvent(String eventName) {
        if (span != null) {
            span.addEvent(eventName);
        }
    }

    void addPreviousScreenAttribute(String screenName) {
        if (span == null) {
            return;
        }
        String previouslyVisibleScreen = visibleScreenTracker.getPreviouslyVisibleScreen();
        if (!screenName.equals(previouslyVisibleScreen)) {
            span.setAttribute(DisneyOtel.LAST_SCREEN_NAME_KEY, previouslyVisibleScreen);
        }
    }
}

File: app/src/main/java/com/raju/disney/opentelemetry/ActiveSpan.kt
package com.raju.disney.opentelemetry

import io.opentelemetry.api.trace.Span
import io.opentelemetry.context.Scope
import java.util.function.Supplier

internal class ActiveSpan(private val visibleScreenTracker: VisibleScreenTracker) {

    private var span: Span? = null
    private var scope: Scope? = null

    fun spanInProgress(): Boolean {
        return span != null
    }

    fun startSpan(spanCreator: Supplier<Span?>) {
        //don't start one if there's already one in progress
        if (span != null) {
            return
        }
        span = spanCreator.get()
        scope = span?.makeCurrent()
    }

    fun endActiveSpan() {
        if (scope != null) {
            scope?.close()
            scope = null
        }
        if (span != null) {
            span?.end()
            span = null
        }
    }

    fun addEvent(eventName: String) {
        span?.addEvent(eventName)
    }

    fun addPreviousScreenAttribute(screenName: String) {
        if (span == null) {
            return
        }
        val previouslyVisibleScreen = visibleScreenTracker.previouslyVisibleScreen
        if (screenName != previouslyVisibleScreen) {
            span?.setAttribute(DisneyOTel.LAST_SCREEN_NAME_KEY, previouslyVisibleScreen)
        }
    }
}